viewing_party/party.py

```diff
@@ -1,23 +1,101 @@
-# ------------- WAVE 1 --------------------
-
 def create_movie(title, genre, rating):
-    pass
+    if title and genre and rating:
+        return {"title": title, "genre": genre, "rating": rating}
+    else: 
+        return None
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
 
-# -----------------------------------------
-# ------------- WAVE 2 --------------------
-# -----------------------------------------
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
 
+def watch_movie(user_data, title):
+    for movie in user_data["watchlist"]:
+        if title == movie["title"]:
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
+            return user_data        
+    return user_data
 
-# -----------------------------------------
-# ------------- WAVE 3 --------------------
-# -----------------------------------------
+def get_watched_avg_rating(user_data):
+    sum = 0.0
+    if len(user_data["watched"]) == 0:
+        return 0.0
+    for movie in user_data["watched"]:
+                sum += movie["rating"]
+    average_rating = sum/len(user_data["watched"])
+    return average_rating
 
+def get_most_watched_genre(user_data):
+    genre_list = []
+    if len(user_data["watched"]) == 0:
+        return None
+    for movie in user_data["watched"]:
+        genre_list.append(movie["genre"])
+    most_genre = max(genre_list, key=genre_list.count)
+    
+    return most_genre
+        
+def get_unique_watched(user_data):
+    user_watched_only = []
+    friends_watched = []
+    unique_list = []
+    for movie in user_data["watched"]:
+        user_watched_only.append(movie)
         
-# -----------------------------------------
-# ------------- WAVE 4 --------------------
-# -----------------------------------------
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+                friends_watched.append(movie)
+                
+    for movie in user_watched_only:
+        if movie not in friends_watched:

```

Comment:
Since we're using the `in` operator here with a list, we will encounter linear time complexity nested within a for loop which would lead us to an overall quadratic time complexity for this method. We could take advantage of constant look up time that `in` has with sets though.

What if user_watched_only and friends_watched were sets with just movie titles as strings in them? Then you could have:

```python
for movie in user_watched_only:
    if movie["title"] not in friends_watched:
        unique_list.append(movie)
return unique_list
```

which would bring us to linear time complexity.